lib/ex_doc/autolink.ex

```diff
@@ -0,0 +1,397 @@
+    :module_id,
+    :id,
+    ext: ".html",
+    skip_undefined_reference_warnings_on: []
+  ]
+
+  alias ExDoc.Formatter.HTML.Templates, as: T
+  alias ExDoc.Refs
+
+  @hexdocs "https://hexdocs.pm/"
+  @otpdocs "http://www.erlang.org/doc/man/"
+
+  def doc(ast, options \\ []) do
+    config = struct!(__MODULE__, options)
+    walk(ast, config)
+  end
+
+  defp walk(list, config) when is_list(list) do
+    Enum.map(list, &walk(&1, config))
+  end
+
+  defp walk(binary, _) when is_binary(binary) do
+    binary
+  end
+
+  defp walk({:pre, _, _} = ast, _config) do
+    ast
+  end
+
+  defp walk({:a, attrs, _} = ast, config) do
+    if url = custom_link(attrs, config) do
+      {:a, Keyword.put(attrs, :href, url), ast}
+    else
+      ast
+    end
+  end
+
+  defp walk({:code, _, [text]} = ast, config) do
+    case text_to_ref(text, :regular, config) do
+      :no_ref ->
+        ast
+
+      ref ->
+        if url = url(ref, config) do
+          {:a, [href: url], [ast]}
+        else
+          ast
+        end
+    end
+  end
+
+  defp walk({tag, attrs, ast}, config) do
+    {tag, attrs, walk(ast, config)}
+  end
+
+  defp custom_link(attrs, config) do
+    with {:ok, href} <- Keyword.fetch(attrs, :href),
+         [[_, text]] <- Regex.scan(~r/^`(.+)`$/, href) do
+      case text_to_ref(text, :custom_link, config) do
+        :no_ref ->
+          nil
+
+        ref ->
+          url(ref, config)
+      end
+    else
+      _ -> nil
+    end
+  end
+
+  defp text_to_ref("c:" <> text, mode, config) do
+    with {call, {:function, module, name, arity}} <- text_to_ref(text, mode, config) do
+      {call, {:callback, module, name, arity}}
+    end
+  end
+
+  defp text_to_ref("t:" <> text, mode, config) do
+    with {call, {:function, module, name, arity}} <- text_to_ref(text, mode, config) do
+      {call, {:type, module, name, arity}}
+    end
+  end
+
+  defp text_to_ref("mix help " <> name, _mode, _config), do: mix_task(name)
+  defp text_to_ref("mix " <> name, _mode, _config), do: mix_task(name)
+
+  defp text_to_ref(text, mode, config) do
+    if not String.contains?(text, [" ", "(", ")"]) do
+      case Code.string_to_quoted(text) do
+        {:ok, {:__aliases__, _, _} = module} ->
+          {:module, module(module)}
+
+        {:ok, {:/, _, [{{:., _, [module, name]}, _, []}, arity]}}
+        when is_atom(name) and is_integer(arity) ->
+          {:remote, {:function, module(module), name, arity}}
+
+        {:ok, {:/, _, [{name, _, _}, arity]}}
+        when is_atom(name) and is_integer(arity) ->
+          {:local, {:function, config.current_module, name, arity}}
+
+        {:ok, erlang_module} when is_atom(erlang_module) and mode == :custom_link ->
+          {:module, erlang_module}
+
+        _ ->
+          :no_ref
+      end
+    else
+      :no_ref
+    end
+  end
+
+  defp module({:__aliases__, _, _} = module), do: Module.concat([Macro.to_string(module)])
+  defp module(module) when is_atom(module), do: module
```

Comment:
@josevalim I previously had this as simply `Macro.expand(module, __ENV__)` - would that be ok? I noticed this crash on Elixir master **when running ex_doc as escript**:

```
** (UndefinedFunctionError) function Mix.Compilers.ApplicationTracer.trace/2 is undefined (module Mix.Compilers.ApplicationTracer is not available)
    Mix.Compilers.ApplicationTracer.trace({:alias_reference, [line: 1], Foo.Bar.Baz}, #Macro.Env<aliases: [{T, ExDoc.Formatter.HTML.Templates}, {Refs, ExDoc.Refs}], context: nil, context_modules: [ExDoc.Autolink], file: "/Users/wojtek/src/ex_doc/lib/ex_doc/autolink.ex", function: {:module, 1}, functions: [{Kernel, [!=: 2, !==: 2, *: 2, ...]}], lexical_tracker: nil, line: 132, macro_aliases: [], macros: [{Kernel, ...}], module: ExDoc.Autolink, requires: [...], ...>)
    (elixir 1.11.0-dev) src/elixir_env.erl:36: :elixir_env."-trace/2-lc$^0/1-0-"/3
    (elixir 1.11.0-dev) src/elixir_env.erl:36: :elixir_env.trace/2
    (elixir 1.11.0-dev) lib/macro.ex:1389: Macro.do_expand_once/2
    (elixir 1.11.0-dev) lib/macro.ex:1559: Macro.expand_until/2
    (ex_doc 0.21.3) lib/ex_doc/autolink.ex:109: ExDoc.Autolink.text_to_ref/3
    (ex_doc 0.21.3) lib/ex_doc/autolink.ex:58: ExDoc.Autolink.walk/2
    (elixir 1.11.0-dev) lib/enum.ex:1396: Enum."-map/2-lists^map/1-0-"/2
```

no such crash when running on Elixir 1.10.2.